Applications/Utils/ModelPreparation/PartitionMesh/NodeWiseMeshPartitioner.cpp

```diff
@@ -0,0 +1,564 @@
+        os_subd_head << " " << partition.regular_elements.size();
+        os_subd_head << " " << partition.ghost_elements.size();
+        os_subd_head << " " << partition.number_of_non_ghost_base_nodes;
+        os_subd_head << " " << partition.number_of_non_ghost_nodes;
+        os_subd_head << " " << _mesh->getNumberOfBaseNodes();
+        os_subd_head << " " << _mesh->getNumberOfNodes();
+        os_subd_head << " " << getNumberOfIntegerVariablesOfElements(
+                                   partition.regular_elements);
+        os_subd_head << " " << getNumberOfIntegerVariablesOfElements(
+                                   partition.ghost_elements)
+                     << " 0 \n";
+    }
+}
+
+void NodeWiseMeshPartitioner::writeElementsASCII(const std::string& file_name_base)
+{
+    const std::string fname = file_name_base + "_partitioned_elems_"
+                              + std::to_string(_npartitions) + ".msh";
+    std::fstream os_subd(fname, std::ios::out | std::ios::trunc);
+    for (const auto& partition : _partitions)
+    {
+        // Set the local node indices of the current partition.
+        IntegerType node_local_id_offset = 0;
+        std::vector<IntegerType> nodes_local_ids(_mesh->getNumberOfNodes(), -1);
+        for (const auto* node : partition.nodes)
+        {
+            nodes_local_ids[node->getID()] = node_local_id_offset;
+            node_local_id_offset++;
+        }
+
+        for (const auto* elem : partition.regular_elements)
+        {
+            writeLocalElementNodeIndices(os_subd, *elem, nodes_local_ids);
+        }
+        for (const auto* elem : partition.ghost_elements)
+        {
+            writeLocalElementNodeIndices(os_subd, *elem, nodes_local_ids);
+        }
+        os_subd << std::endl;
+    }
+}
+
+void NodeWiseMeshPartitioner::writeNodesASCII(const std::string& file_name_base)
+{
+    const std::string fname = file_name_base + "_partitioned_nodes_"
+                              + std::to_string(_npartitions) + ".msh";
+    std::fstream os_subd_node(fname, std::ios::out | std::ios::trunc);
+    os_subd_node.precision(std::numeric_limits<double>::digits10);
+    os_subd_node.setf(std::ios::scientific);
+
+    for (const auto& partition : _partitions)
+    {
+        for (const auto* node : partition.nodes)
+        {
+            double const* coords = node->getCoords();
+            os_subd_node << _nodes_global_ids[node->getID()] << " " << coords[0]
+                         << " " << coords[1] << " " << coords[2] << "\n";
+        }
+        os_subd_node << std::endl;
+    }
+}
+
+void NodeWiseMeshPartitioner::writeASCII(const std::string& file_name_base)
+{
+    writeConfigDataASCII(file_name_base);
+    writeElementsASCII(file_name_base);
+    writeNodesASCII(file_name_base);
+}
+
+void NodeWiseMeshPartitioner::resetGlobalNodeIndices()
+{
+    for (std::size_t i = 0; i < _mesh->_nodes.size(); i++)
+    {
+        _mesh->_nodes[i]->setID(_nodes_global_ids[i]);
```

Comment:
Since the element nodes are pointers to Node, there is no need to process elements.